### PR TITLE
Comp ships

### PR DESCRIPTION
**Pull request by nagerz**
@aprildagonese Check this out. Got all ships to place repeatably. No issues from what I can tell after many runs. Cleanest code yet for me I think. Broke out extra methods for you. See if you can follow/understand before I explain it.

### CODE REVIEW

File: lib/computer_brain.rb [hunk 1]
```diff
@@ -1,14 +1,101 @@
...
+    end
+    return potential_first_keys
+  end
+
+  def directional_search(coordinate, direction, ship)
+    letter = coordinate.split("")[0]
+    number = coordinate.split("")[1].to_i
+    potential_keys = []
+
+    if direction == :vertical
```
Comment (aprildagonese) [inline lib/computer_brain.rb]: Do you think it makes sense to put all of the 'if direction == :vertical' and the 'if direction == :horizontal' code into separate methods for readability? That's what everyone kept telling me to do yesterday when I was getting help on the get_last_hit problem.

File: lib/computer_brain.rb [hunk 2]
```diff
@@ -1,14 +1,101 @@
...
+      test_range = (letter..end_letter).to_a
+      test_range.each do |dyn_letter|
+        potential_keys << "#{dyn_letter}#{number}"
+      end
+      if valid_potential_keys?(potential_keys) == true
+        return potential_keys
+      else
+        return false
+      end
+    elsif direction == :horizontal
```
Comment (aprildagonese) [inline lib/computer_brain.rb]: Same for this section?

File: lib/computer_brain.rb [hunk 3]
```diff
@@ -1,14 +1,101 @@
...
+      end_number = number + ship.length - 1
+      test_range = (number..end_number).to_a
+      test_range.each do |dyn_number|
+        potential_keys << "#{letter}#{dyn_number}"
+      end
+      if valid_potential_keys?(potential_keys) == true
```
Comment (aprildagonese) [inline lib/computer_brain.rb]: Also, seeing some repetition between lines 41 and 52, which might a chance for another small method!

File: lib/computer_brain.rb [hunk 4]
```diff
@@ -1,14 +1,101 @@
...
+        return potential_keys
+      else
+        return false
+      end
+    end
+  end
+
+  def valid_potential_keys?(potential_keys)
```
Comment (aprildagonese) [inline lib/computer_brain.rb]: Love.

File: lib/computer_brain.rb [hunk 5]
```diff
@@ -1,14 +1,101 @@
...
+
+  def cpu_place_ships(ships)
+    valid_ship_counter = 0
+    @validated_keys = []
+    validated_ships = []
+    ships.each do |ship|
+      potential_array = make_limited_placement_array(ship)
+      first_key = potential_array.sample
+      if !@validated_keys.include?(first_key)
+        search_direction = [:horizontal, :vertical].sample
```
Comment (aprildagonese) [inline lib/computer_brain.rb]: This is cool!

File: lib/computer_brain.rb [hunk 6]
```diff
@@ -1,14 +1,101 @@
...
+    end
+  end
+
+  def valid_potential_keys?(potential_keys)
+    potential_keys.none? do |key|
+      @validated_keys.include?(key)
+    end
+  end
+
+  def cpu_place_ships(ships)
```
Comment (aprildagonese) [inline lib/computer_brain.rb]: This guy is a beast! Your logic here is super cool, although the method itself is pretty heavy. Any chances to break this one down into smaller pieces? Also, I ran into needing to know the available ships on my side, too. Maybe we should turn to that next?

File: lib/computer_brain.rb (hunk 2 above)
Comment (nagerz) [reply lib/computer_brain.rb]: Will look at splitting it out.

File: lib/computer_brain.rb (hunk 3 above)
Comment (nagerz) [reply lib/computer_brain.rb]: Those lines are the same but they are already calling another method. You could remove the "== true" and it wouldn't look as repetitive but I like those for clarity.

File: lib/computer_brain.rb (hunk 4 above)
Comment (nagerz) [reply lib/computer_brain.rb]: Thanks

File: lib/computer_brain.rb (hunk 6 above)
Comment (nagerz) [reply lib/computer_brain.rb]: We can look at it together

**Pull request merged**